js/modules/types/message.js

```diff
@@ -1,17 +1,165 @@
+const isFunction = require('lodash/isFunction');
+
 const Attachment = require('./attachment');
+const Errors = require('./errors');
+const SchemaVersion = require('./schema_version');
 
 
 const GROUP = 'group';
 const PRIVATE = 'private';
 
+// Schema version history
+//
+// Version 0
+//   - Schema initialized
+// Version 1
+//   - Attachments: Auto-orient JPEG attachments using EXIF `Orientation` data
+// Version 2
+//   - Attachments: Sanitize Unicode order override characters
+const INITIAL_SCHEMA_VERSION = 0;
+
+// Increment this version number every time we add a message schema upgrade
+// step. This will allow us to retroactively upgrade existing messages. As we
+// add more upgrade steps, we could design a pipeline that does this
+// incrementally, e.g. from version 0 / unknown -> 1, 1 --> 2, etc., similar to
+// how we do database migrations:
+exports.CURRENT_SCHEMA_VERSION = 2;
```

Comment:
One thing I've gone back and forth on with the iOS approach is whether it would be easier to have a single schema version for the entire database rather than a per model one like we do.

The consequences are a bit subtle, and mostly affect us around inheritance (e.g. migrations in a superclass and some migrations in it's subclass), but if I could do it all over again, I think I'd have one for the entire db and have the "upgrade" process built into our base ORM class. 

This might not be very relevant to Desktop land at all - I don't even know if you use inheritance, but nice to avoid copy/pasting slightly different versions of this machinery around.

Reply:
@michaelkirk-signal This is absolutely relevant to us as well and I appreciate your feedback. We use IndexedDB (IDB) which does have a built-in global version number. We are currently at 16: https://github.com/signalapp/Signal-Desktop/blob/bbbd52798523770be28c65af9c06065b4bc40886/js/database.js#L227-L235

We use the Backbone framework to implement our models and while it normally works with Ajax to save information to the server we use a ‘sync adapter’ for IDB to save information locally. Our IDB adapter has a built-in migration mechanism: https://github.com/superfeedr/indexeddb-backbonejs-adapter/blob/446e25040ff0e5e6e9404893cfdb41984dce5d67/backbone-indexeddb.js#L131-L191

Unfortunately, it is limited as transactions during the IDB upgrade cannot do async operations that return control flow to the event loop, auto-rotate JPEGS or generate cryptographic hash, etc., without the transaction going inactive.

At the same time, we are balancing maintenance vs UX. For easier maintenance it’d be better to upgrade all database entries at once, e.g. during startup, but that would make the user wait (bad UX). For now, we’ve decided to do structural changes to the database, e.g. adding stores (like tables) and indexes, during startup and heavier-weight data migrations, e.g. auto-rotate JPEGs, re-computing attachment digests, on a lazy basis similar to what you described you do for iOS.